feat: [HL-23] Create account screen layout 🍕

abdelhamid-f-nasser opened the pull request:
<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
  
  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.
  
  Before submitting a Pull Request, please ensure you've done the following:
  - 📖 Read the Open Sauced Contributing Guide: https://github.com/open-sauced/open-sauced/blob/HEAD/CONTRIBUTING.md#create-a-pull-request.
  - 📖 Read the Open Sauced Code of Conduct: https://github.com/open-sauced/open-sauced/blob/HEAD/CODE_OF_CONDUCT.md.
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - ✅ Provide tests for your changes.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
-->

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
## Description
Adds the Create Account screen layout
<!-- 
Please do not leave this blank 
This PR [adds/removes/fixes/replaces] the [feature/bug/etc]. 
-->
## Related Tickets & Documents
<!-- 
Please use this format link issue numbers: Fixes #123
https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword 
-->
## Added tests?

- [x] 🙋 no, because I need help

## Added to documentation?

- [x] 🙅 no documentation needed

## [optional] Are there any post-deployment tasks we need to perform?



## [optional] What gif best describes this PR or how it makes you feel?

abdelhamid-f-nasser committed:
Add checkbox asset files  🖼️

- Add checkbox assets

Note: Selected is png not pdf (asset unavailable)

abdelhamid-f-nasser committed:
Create CheckboxButton.swift

Add checkbox button

abdelhamid-f-nasser committed:
Add CreateAccountViewController ✨

abdelhamid-f-nasser committed:
Merge branch 'develop' of https://github.com/motoon-eg/healthy into feat/HL-23-Create-account-screen-layout

* 'develop' of https://github.com/motoon-eg/healthy:
  feat: [HL-1] Configure Generating Localizable Strings (#16)

abdelhamid-f-nasser committed:
Resolve lint warnings 🔧

- Fix File Header Violation in `CreateAccountViewController.swift` line 5
- Fix line length Violation in `CheckboxButton.swift` line 21
- Fix line length Violation in `CreateAccountViewController.swift` line 29

abdallah-7698 commented on Healthy/Classes/Modules/CreateAccount/CreateAccountViewController.swift:
```diff
@@ -1,37 +1,19 @@
     @IBOutlet weak var passwordTextField: UITextField!
     @IBOutlet weak var confirmPasswordTextField: UITextField!
     @IBOutlet weak var checkBoxButton: CheckboxButton!
     @IBOutlet weak var signInTextLabel: UILabel!
```
@IBOutlet weak private var nameTextField: UITextField!
    @IBOutlet weak private var emailTextField: UITextField!
    @IBOutlet weak private var passwordTextField: UITextField!
    @IBOutlet weak private var confirmPasswordTextField: UITextField!
    @IBOutlet weak private var checkBoxButton: CheckboxButton!
    @IBOutlet weak private var signInTextLabel: UILabel!

abdallah-7698 commented on Healthy/Classes/ReusableViews/CheckboxButton.swift:
```diff
@@ -0,0 +1,30 @@
+    // Bool property
+    var isChecked: Bool = false {
+        didSet {
+            if isChecked == true {
+                self.setImage(checkedImage, for: UIControl.State.normal)
+            } else {
+                self.setImage(uncheckedImage, for: UIControl.State.normal)
```
self.setImage(isChecked ? .icCheckboxSelected : .icCheckboxNotSelected, for: UIControl.State.normal)

abdelhamid-f-nasser replied:
Done, thanks for reviewing !! However the line length is 100 😭

abdelhamid-f-nasser replied:
I've wrote it as 
```swift
let image = isChecked ? UIImage.icCheckboxSelected : UIImage.icCheckboxNotSelected
self.setImage(image, for: UIControl.State.normal)
```

abdallah-7698 commented on Healthy/Classes/ReusableViews/CheckboxButton.swift:
```diff
@@ -0,0 +1,30 @@
+
+    // Images
+    let checkedImage = UIImage.icCheckboxSelected
+    let uncheckedImage = UIImage.icCheckboxNotSelected
```
based on above remove this

abdelhamid-f-nasser replied:
Done ! 👍🏼

abdallah-7698 commented on Healthy/Classes/ReusableViews/CheckboxButton.swift:
```diff
@@ -0,0 +1,30 @@
+        self.isChecked = false
+    }
+
+    @objc func buttonClicked(sender: UIButton) {
+        if sender == self {
+            isChecked = !isChecked
+        }
+    }
```
@objc private func buttonClicked(sender: UIButton) {
            isChecked = !isChecked
    }
if you set the method to private you don't need to check the sender

abdelhamid-f-nasser replied:
Done !  👍🏼

oradwanomar commented on Healthy/Classes/Modules/CreateAccount/CreateAccountViewController.swift:
```diff
@@ -0,0 +1,19 @@
+
+class CreateAccountViewController: UIViewController {
+
+    @IBOutlet weak var nameTextField: UITextField!
+    @IBOutlet weak var emailTextField: UITextField!
+    @IBOutlet weak var passwordTextField: UITextField!
+    @IBOutlet weak var confirmPasswordTextField: UITextField!
+    @IBOutlet weak var checkBoxButton: CheckboxButton!
+    @IBOutlet weak var signInTextLabel: UILabel!
```
```suggestion
    @IBOutlet private (set) weak var nameTextField: UITextField!
    @IBOutlet private (set) weak var emailTextField: UITextField!
    @IBOutlet private (set) weak var passwordTextField: UITextField!
    @IBOutlet private (set) weak var confirmPasswordTextField: UITextField!
    @IBOutlet private (set) weak var checkBoxButton: CheckboxButton!
    @IBOutlet private (set) weak var signInTextLabel: UILabel!
```

abdelhamid-f-nasser replied:
Done ! Thank you for reviewing ^ ^

oradwanomar reviewed (commented):
I like the way you write the PR description ❤️❤️🧨 
Keep going

ahmdmhasn committed:
Merge remote-tracking branch 'origin/develop' into feat/HL-23-Create-account-screen-layout

# Conflicts:
#	Healthy.xcodeproj/project.pbxproj

abdelhamid-f-nasser committed:
Adjust layout to decrease use of constraints

Adjust spacing and limit use of constraints

abdelhamid-f-nasser committed:
Refactor(CheckboxButton): change checkedImage, uncheckedImage to private

abdelhamid-f-nasser committed:
Merge branch 'feat/HL-23-Create-account-screen-layout' of https://github.com/motoon-eg/healthy into feat/HL-23-Create-account-screen-layout

* 'feat/HL-23-Create-account-screen-layout' of https://github.com/motoon-eg/healthy:
  feat: [HL-21] Create Login view + layout ✨ (#23)
  feat: [HL-5] Update button styles (#20)

abdelhamid-f-nasser committed:
Change access modifier of IBOutlet in signUpButton to private

Update the access modifiers of IBOutlets to hide the setter from external classes

abdelhamid-f-nasser committed:
refactor(Checkbox): Use ternary operator for button status checks

- Use ternary operator for checking the current button statusRemove
- Remove redundant properties

abdelhamid-f-nasser committed:
refactor(Checkbox): encapsulate buttonClicked() method

- Changed buttonClicked to private for better privacy and encapsulation
- Remove redundant check for sender == self

abdelhamid-f-nasser committed:
style: Reformat the code

abdelhamid-f-nasser committed:
revert(Assets): Remove the icons from the directory

abdelhamid-f-nasser committed:
style: Reformat the code

ahmdmhasn commented on Healthy/Resources/Assets.xcassets/ic_checkbox_selected.imageset/ic_checkbox_selected-1x.png:
Let's use PDF instead of PNG

abdelhamid-f-nasser replied:
I didn't find this asset in form of a PDF file, Should I convert it to a PDF file or should i search for another asset?

ahmdmhasn commented on Healthy/Classes/ReusableViews/CheckboxButton.swift:
```diff
@@ -0,0 +1,21 @@
+import UIKit
+
+class CheckboxButton: UIButton {
+
+    // Bool property
+    var isChecked: Bool = false {
+        didSet {
+            let image = isChecked ? UIImage.icCheckboxSelected : UIImage.icCheckboxNotSelected
+            self.setImage(image, for: UIControl.State.normal)
+        }
+    }
+
+    override func awakeFromNib() {
+        self.addTarget(self, action: #selector(buttonClicked(sender:)), for: .touchUpInside)
+        self.isChecked = false
+    }
+
+    @objc private func buttonClicked(sender: UIButton) {
+        isChecked = !isChecked
+    }
+}
```
```suggestion
import UIKit

final class CheckboxButton: UIButton {

    // MARK: Properties

    private(set) var isChecked: Bool = false {
        didSet {
            let image = isChecked ? UIImage.icCheckboxSelected : UIImage.icCheckboxNotSelected
            self.setImage(image, for: UIControl.State.normal)
        }
    }

    // MARK: Lifecycle

    override func awakeFromNib() {
        self.addTarget(self, action: #selector(buttonClicked(sender:)), for: .touchUpInside)
        self.isChecked = false
    }

    // MARK: Actions

    @objc private func buttonClicked(sender: UIButton) {
        isChecked.toggle()
    }
}
```

abdelhamid-f-nasser replied:
Done, thanks I'll make sure to always follow this !! 👍🏼

ahmdmhasn commented on Healthy/Classes/Modules/CreateAccount/CreateAccountViewController.swift:
```diff
@@ -0,0 +1,30 @@
+import UIKit
+
+class CreateAccountViewController: UIViewController {
+
+    @IBOutlet private (set) weak var parentVerticalStackView: UIStackView!
+
+    @IBOutlet private (set) weak var nameTextField: UITextField!
+
+    @IBOutlet private (set) weak var emailTextField: UITextField!
+
+    @IBOutlet private (set) weak var passwordTextField: UITextField!
+
+    @IBOutlet private (set) weak var confirmPasswordContainerView: UIView!
+    @IBOutlet weak var confirmPasswordTextField: UITextField!
+
+    @IBOutlet private (set) weak var termsAndConditionsHorizontalStackView: UIStackView!
+    @IBOutlet private (set) weak var checkBoxButton: CheckboxButton!
+    @IBOutlet private (set) weak var acceptTermsTextLabel: UILabel!
+
+    @IBOutlet private (set) weak var signInTextLabel: UILabel!
+    @IBOutlet private (set) weak var signUpButton: UIButton!
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+    }
+
+    @IBAction func didTapSignUp(_ sender: Any) {
+    }
+
+}
```
```suggestion
import UIKit

final class CreateAccountViewController: UIViewController {

    // MARK: Outlets

    @IBOutlet private (set) weak var parentVerticalStackView: UIStackView!
    @IBOutlet private (set) weak var nameTextField: UITextField!
    @IBOutlet private (set) weak var emailTextField: UITextField!
    @IBOutlet private (set) weak var passwordTextField: UITextField!
    @IBOutlet private (set) weak var confirmPasswordContainerView: UIView!
    @IBOutlet private (set) weak var confirmPasswordTextField: UITextField!
    @IBOutlet private (set) weak var termsAndConditionsHorizontalStackView: UIStackView!
    @IBOutlet private (set) weak var checkBoxButton: CheckboxButton!
    @IBOutlet private (set) weak var acceptTermsTextLabel: UILabel!
    @IBOutlet private (set) weak var signInTextLabel: UILabel!
    @IBOutlet private (set) weak var signUpButton: UIButton!
    
    // MARK: Lifecycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
}

// MARK: Actions

private extension CreateAccountViewController { 
    @IBAction func didTapSignUp(_ sender: Any) {
    }
}
```

ahmdmhasn replied:
Consistency is the key!

ahmdmhasn commented on Healthy/Classes/Generated/UIImage.Generated.swift:
```diff
@@ -9,4 +9,16 @@ extension UIImage {
+    }
+
+    static var icCheckboxNotSelected: UIImage {
+      UIImage(named: "ic_checkbox_not_selected")!
+    }
+
+    static var icCheckboxSelected: UIImage {
+      UIImage(named: "ic_checkbox_selected")!
```
```suggestion
      UIImage(named: "icon_checkbox_not_selected")!
    }

    static var icCheckboxSelected: UIImage {
      UIImage(named: "icon_checkbox_selected")!
```

This is more a Swifty way... But thanks for keeping a convention when picking a name for the assets.

ahmdmhasn commented:
@abdelhamid-f-nasser Hello hello 👀

abdelhamid-f-nasser committed:
Merge branch 'develop' of https://github.com/motoon-eg/healthy into feat/HL-23-Create-account-screen-layout

* 'develop' of https://github.com/motoon-eg/healthy:
  feat: [HL-22] add validator with the ability to validate email & password✌️ (#37)
  feat: [HL-36] Create logger with default logger🍕 (#36)
  build: Ignore unmodified files in SwiftLint script 🔨  (#42)
  feat: [HL-19] Add and configure FormTextField (#19)
  fix: Colors not added to target files (#43)
  chore: Auto generate colors via SwiftGen 🎨 (#35)
  revert: feat: [HL-25] add validator with the ability to validate email & password✌️
  revert: Missing comments in splash screen 👨🏼‍🔧 (#40)
  feat: [HL-20] Create sign button with social media view 📖 (#25)
  feat: [HL-3] Add splash screen 🍕 (#38)
  feat: [HL-18] Add and configure Xcode template 🔨 (#22)

# Conflicts:
#	Healthy.xcodeproj/project.pbxproj
#	Healthy/Classes/Generated/UIImage.Generated.swift
#	Healthy/Classes/Modules/Login/LoginViewController.swift

ahmdmhasn commented on Healthy.xcodeproj/project.xcworkspace/xcshareddata/WorkspaceSettings.xcsettings:
Please don't forget to remove this file

abdelhamid-f-nasser committed:
refactor: Rename checkbox icons to icon-checkbox

- Change unselected icon name from ic_checkbox_not_selected to icon-checkbox-not-selected
- Change selected icon name from ic_checkbox_selected to icon-checkbox-selected

abdelhamid-f-nasser committed:
chore: Remove `WorkspaceSettings.xcsettings`

abdelhamid-f-nasser committed:
style: Add mark comments for better readability and organization

abdelhamid-f-nasser committed:
📝 docs: Add documentation for `Checkbox.swift`

abdelhamid-f-nasser committed:
style: Resolve lint warnings 🧹
